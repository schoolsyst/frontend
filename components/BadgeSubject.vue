<template>
  <div
    :class="{
      '--badge-subject': true,
      expandable,
      expanded,
      editable,
    }"
    :style="{ backgroundColor: bgColor, color: fgColor }"
    :title="badgeTitle"
    @mouseenter="expanded = alwaysExpanded || expandable"
    @mouseleave="expanded = alwaysExpanded"
  >
    <button-icon small class="button-edit" @click="$emit('edit')"
      >edit</button-icon
    >
    <div class="name">
      <slot></slot>
    </div>
  </div>
</template>

<script lang="ts">
// eslint-disable-next-line no-unused-vars
import Vue, { PropOptions } from 'vue'
import { isHexColor } from '~/utils/validators'
import { lightColor, darkColor } from '~/utils/colors'
import ButtonIcon from '~/components/ButtonIcon.vue'

export default Vue.extend({
  components: { ButtonIcon },
  props: {
    expandable: {
      type: Boolean,
      default: false,
    } as PropOptions<boolean>,
    alwaysExpanded: {
      type: Boolean,
      default: false,
    } as PropOptions<boolean>,
    color: {
      type: String,
      required: true,
      validator: isHexColor,
    } as PropOptions<string>,
    editable: {
      type: Boolean,
      default: false,
    } as PropOptions<boolean>,
  },
  data() {
    return { expanded: this.alwaysExpanded }
  },
  computed: {
    bgColor(): string {
      return this.expanded ? lightColor(this.color) : this.color
    },
    fgColor(): string {
      return darkColor(this.color)
    },
    badgeTitle(): string | null {
      const slotText = this.$slots.default?.[0]?.text
      if (slotText && !this.expanded) {
        return slotText
      }
      return null
    },
  },
})
</script>

<style lang="stylus" scoped>
//
//Definitions
//
dot-size = 1.4em

//
//Positioning
//
.--badge-subject
  display inline-flex
  align-items center

.--badge-subject.editable .name
  margin-left 0.5em

.name
  white-space nowrap

.--badge-subject:not(.editable) .button-edit
  width 0

//
//Sizing
//
.--badge-subject:not(.expanded)
  max-width dot-size

//
//Spacing
//
.button-edit
  width dot-size
  font-size 0.85em

.--badge-subject
  padding 0 0.5em
  height dot-size

//
//Color
//

//
//Decoration
//
.--badge-subject
  border-radius 1000px

.--badge-subject:not(.editable) .button-edit
  cursor default

//
//Visibility
//
.name
  overflow hidden
  opacity 0

.button-edit
  opacity 0

//
//States
//

//When the element is not expanded and _can_ be expanded
.expandable
  transition 250ms ease all

.expanded.editable

  .button-edit
    opacity 1

.expanded
  max-width 25ch

  .name
    opacity 1
</style>
