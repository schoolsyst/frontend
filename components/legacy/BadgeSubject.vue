<template lang="pug">
component.subject(
  :is="element"
  :style="{backgroundColor, color: subjectNameColor}"
  :class="{clickable, multiline, thin, inline}"
  :data-variant="variant"
  type="button"
  @click="$emit('click')"
  v-tooltip="tooltipContent"
)
  Icon(v-if="!color").unknown-icon more_horiz
  span.name {{ name || placeholderName }}
</template>

<script>
import { mapGetters } from 'vuex'
import Icon from '~/components/legacy/Icon.vue'
export default {
  name: 'BadgeSubject',
  components: { Icon },
  props: {
    // Data
    color: {
      type: String,
      default: null,
    },
    name: {
      type: String,
      default: null,
    },
    // Supporting data
    placeholderName: {
      type: String,
      default: 'Choisir...',
    },
    // Style
    clickable: {
      type: Boolean,
      default: false,
    },
    multiline: {
      type: Boolean,
      default: false,
    },
    variant: {
      type: String,
      default: 'badge',
    },
    thin: {
      type: Boolean,
      default: false,
    },
    inline: {
      type: Boolean,
      default: false,
    },
    noTooltip: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    backgroundColor() {
      return this.color || 'var(--black)'
    },
    subjectNameColor() {
      return this.color ? this.textColor()(this.color) : 'var(--white)'
    },
    element() {
      return this.clickable ? 'button' : 'span'
    },
    tooltipContent() {
      if (this.noTooltip) return null
      return this.name || this.placeholderName
    },
  },
  methods: {
    ...mapGetters(['textColor']),
  },
}
</script>

<style lang="stylus" scoped>
//
//Definitions
//
badge-aspect()
  //==== Positioning
  align-items center
  //==== Spacing
  padding 0.7em
  //==== Decorations
  border-radius var(--border-radius)

dot-aspect()
  //==== Definitions
  size = 1.2em

  //==== Positioning
  display flex
  flex-shrink 0
  justify-content center
  align-items center
  min-width size / 2
  min-height size / 2
  width size
  //==== Sizing
  height size
  //==== Decoration
  border-radius 50%

  //==== Appearance
  .name
    display none

//
//Layout
//
.subject
  position relative
  //==== Positioning
  display block
  //==== Appearance
  overflow hidden
  //==== Sizing
  max-height 25vh

//
//Decoration
//
.subject
  appearance none

//
//Colors
//
.unknown-icon
  color var(--white)

//
//Typography
//
.subject
  text-overflow ellipsis
  font-weight normal

  &:not(.multiline)
    flex-shrink 0
    white-space nowrap

  &.multiline
    white-space wrap

//
//States
//
[data-variant=badge]
  badge-aspect()

  &.thin
    padding 0.3em 0.5em

[data-variant=dot]
  dot-aspect()

[data-variant=pill]
  badge-aspect()
  size = 1.5em

  display flex
  flex-shrink 1
  align-items center
  height size
  border-radius size

[data-variant=responsive]
  @media (max-width: 650px)
    badge-aspect()

  @media (min-width: 651px)
    dot-aspect()

.inline
  display inline-flex
  padding 0
  padding-bottom -0.5em
  width 1em
  height 1em

.clickable
  &:focus, &:hover
    outline none
    opacity 0.75
</style>
