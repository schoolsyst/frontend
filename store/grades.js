import { firstBy } from 'thenby'
import {
  isBefore,
  isWithinInterval,
  parseISO,
  isWeekend,
  differenceInWeeks,
  isSameWeek,
} from 'date-fns'
import { getMutations, getValidator } from './index'

export const state = () => ({
  grades: [],
})

const parseGradeDates = grade => ({
  ...grade,
  added: parseISO(grade.added),
  obtained_date: parseISO(grade.obtained),
})

export const getters = {
  all: (state, getters) => getters.order(state.grades),
  one: (state, getters) => (value, prop = 'uuid') =>
    getters.all.find(o => o[prop] === value) || null,
  of: (state, getters, rootState, rootGetters) => (
    value,
    what = 'trimester',
    date = 'added'
  ) => {
    switch (what) {
      case 'trimester': {
        let trimesterInterval
        if (value === null) {
          trimesterInterval = rootGetters['schedule/currentTrimester']
        } else {
          if (![1, 2, 3].includes(value)) return []
          trimesterInterval = rootGetters['schedule/trimester'](value)
        }
        if (trimesterInterval === null) return []
        // Gets all the grades that have been added within the `value`th trimester
        return getters.all.filter(
          o => o[date] && isWithinInterval(o[date], trimesterInterval)
        )
      }
      case 'interval':
        return getters.all.filter(
          o => o[date] && isWithinInterval(o[date], value)
        )

      default:
        return []
    }
  },
  pending: (state, getters) => getters.all.filter(o => o.obtained === null),
  order: (state, getters) => grades => {
    // ↓ Makes a copy of grades to avoid mutating
    return [...grades].sort(firstBy((o1, o2) => isBefore(o1, o2)))
  },
  format: (state, getters, rootState, rootGetters) => value => {
    let unit = rootGetters['settings/value']('grade_max')
    unit = unit === 100 ? '%' : `/${unit}`
    return { value, unit }
  },
  mean: (state, getters) => (grades, debug = false) => {
    let ret
    grades = grades.filter(o => o.obtained !== null)
    if (!grades.length) ret = null
    else {
      const sumOfWeights = grades
        .map(o => o.weight * o.subject.weight)
        .reduce((acc, cur) => acc + cur)
      const sumOfGrades = grades
        .map(o => o.obtained * o.weight * o.subject.weight)
        .reduce((acc, cur) => acc + cur)

      ret = sumOfGrades / sumOfWeights
    }
    return ret
  },
  evolution: (state, getters) => grades => {
    /* Returns a number ∈ [0,1] that represents the relative gap between...
     * - the old mean (mean before the latest grade), and
     * - the new mean (mean with the latest grade)
     *
     * Effectively represents the relative evolution of the mean caused by the latest grade.
     */
    // Just to be sure, sort the grades
    // Also remove notes that have no `obtained` score (see `const oldMean = `) for why
    grades = getters.order(grades).filter(o => !!o.obtained)
    if (!grades.length) return null
    // Get the current mean (w/ all the grades given)
    const newMean = getters.mean(grades)
    // Get the mean w/o the latest grade *that has been obtained*
    //                                     ↓ we use .slice, .pop would mutate *the state*, which… no.
    const oldMean = getters.mean(grades.slice(0, -1))
    // Classic relative gap formula
    return (newMean - oldMean) / oldMean
  },
  currentTrimester: (state, getters, rootState, rootGetters) => {
    return getters.of(rootGetters['schedule/currentTrimester'], 'trimester')
  },
  currentTrimesterMean: (state, getters, rootState, rootGetters) =>
    getters.mean(getters.currentTrimester),
  currentTrimesterEvolution: (state, getters, rootState, rootGetters) =>
    getters.evolution(getters.currentTrimester),
  display: (_, __, ___, rootGetters) => (
    grade,
    unit = null,
    precision = 2,
    abs = false
  ) => {
    if (!grade && grade !== 0) return '—'
    grade = abs ? ~~grade : grade
    unit = unit || rootGetters['settings/value']('grade_max')
    return (grade * unit).toFixed(precision).replace('.', ',')
  },
  absoluteUnit: (_, __, ___, rootGetters) => (value, unit = null) => {
    unit = unit || rootGetters['settings/value']('grade_max')
    return value / unit
  },
  formatUnit: (_, __, ___, rootGetters) => (value, unit = null) => {
    unit = unit || rootGetters['settings/value']('grade_max')
    return value * unit
  },
  meanOfDays: (state, getters, rootState, rootGetters) => (
    start = null,
    end = null
  ) => {
    let grades
    if (start && end) {
      grades = getters.of({ start, end }, 'interval', 'obtained_date')
    } else if (end === null) {
      grades = getters.of(start, 'trimester', 'obtained_date')
    } else {
      grades = getters.of(null, 'trimester', 'obtained_date')
    }
    return getters.mean(grades)
  },
  ofWeekSmart: (_, { all }) =>
    all.filter(o => {
      const now = Date.now()
      if (isWeekend(now)) return differenceInWeeks(o.added, now) === 1
      else return isSameWeek(o.added, now)
    }),
  validate: getValidator({
    constraints: {
      maxLength: {
        300: ['name'],
      },
      minimum: {
        0: ['obtained', 'expected', 'goal', 'weight'],
        1: ['unit'],
      },
      required: ['name', 'subject', 'weight', 'unit'],
    },
    /* We use custom constraints because the default error message
     * is "should be less than 1", which does not coincide
     * with the UI, which shows the "real", non-normalized grade
     */
    customConstraints: [
      {
        message: 'La note obtenue est trop grande',
        field: 'obtained',
        constraint: (getters, object) =>
          object.obtained === null || object.obtained <= 1,
      },
      {
        message: 'La note estimée est trop grande',
        field: 'expected',
        constraint: (getters, object) =>
          object.expected === null || object.expected <= 1,
      },
      {
        message: "L'objectif de note est trop grand",
        field: 'goal',
        constraint: (getters, object) =>
          object.goal === null || object.goal <= 1,
      },
    ],
    fieldNames: {
      name: { gender: 'M', name: 'nom' },
      obtained: { gender: 'F', name: 'note obtenue' },
      expected: { gender: 'F', name: 'note estimée' },
      goal: { gender: 'M', name: 'objectif de note' },
      weight: { gender: 'M', name: 'coefficient' },
      unit: { gender: 'F', name: 'unité' },
      subject: { gender: 'F', name: 'matière' },
    },
    resourceName: { gender: 'F', name: 'note' },
  }),
}

export const mutations = {
  ...getMutations('grade', parseGradeDates),
}

export const actions = {
  async load({ commit, state }, force = false) {
    if (!force && state.grades.length) return
    try {
      const { data } = await this.$axios.get(`/grades/`)
      // console.log(`[from API] GET /grades/: OK`)
      if (data) commit('SET', data)
    } catch (error) {
      // console.error(`[from API] GET /grades/: Error`)
      try {
        // console.error(error.response.data)
      } catch (_) {
        // eslint-disable-next-line
        console.error(error)
      }
    }
  },
  async post({ commit, getters }, { grade, force }) {
    force = force || false
    if (!force) {
      const validation = getters.validate(grade)
      if (!validation.validated) return false
    }
    try {
      grade.subject = grade.subject.uuid
      const res = await this.$axios.post(`/grades/`, grade)
      const { data } = this.$axios.get(`/grades/${res.data.uuid}`)
      if (data) commit('ADD', data)
      return true
    } catch (error) {
      // eslint-disable-next-line
      console.error(error)
      return false
    }
  },

  async patch(
    { commit, dispatch, getters },
    uuid,
    modifications,
    force = false
  ) {
    if (!force) {
      let grade = getters.one(uuid)
      grade = { ...grade, ...modifications }
      const validation = await dispatch('validate', grade)
      if (!validation.validated) return validation
    }
    try {
      const { data } = await this.$axios.patch(`/grades/${uuid}`, modifications)
      if (data) commit('PATCH', uuid, data)
      // console.log(`[from API] PATCH /grades/${uuid}: OK`)
    } catch (error) {
      // console.error(`[from API] PATCH /grades/${uuid}: Error`)
      try {
        // console.error(error.response.data)
      } catch (_) {
        // eslint-disable-next-line
        console.error(error)
      }
    }
  },
  async delete({ commit }, uuid) {
    try {
      const { data } = await this.$axios.delete(`/grades/${uuid}`)
      if (data) commit('DEL', uuid)
      // console.log(`[from API] DELETE /grades/${uuid}: OK`)
    } catch (error) {
      // console.error(`[from API] DELETE /grades/${uuid}: Error`)
      try {
        // console.error(error.response.data)
      } catch (_) {
        // eslint-disable-next-line
        console.error(error)
      }
    }
  },
}
