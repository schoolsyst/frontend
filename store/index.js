import { toDate, addDays, isBefore, format } from 'date-fns'
import tinycolor from 'tinycolor2'
import constantCase from 'constant-case'
import Vue from 'vue'
const version = require('~/package.json').version.split('.')

// eslint-disable-next-line import/order
const commitHash = require('child_process')
  .execSync('git rev-parse HEAD')
  .toString()
  .trim()
const commitHashShort = commitHash.slice(0, 7)

export const state = () => ({
  version: {
    feature: version[0],
    ui: version[1],
    bug: version[2],
    channel: 'beta',
    build: {
      full: commitHash,
      short: commitHashShort,
    },
  },
  now: toDate(Date.now()), // For time-dependent getters.
  tomorrow: addDays(toDate(Date.now()), 1),
  today: new Date(),
  location: {
    latitude: null,
    longitude: null,
  },
  links: [
    {
      name: 'Timeline',
      href: '/timeline',
      icon: 'timeline',
      id: 'timeline',
    },
    {
      name: 'Cours',
      href: '/notes',
      icon: 'insert_drive_file',
      id: 'notes',
    },
    {
      name: 'Devoirs',
      href: '/homework',
      icon: 'book',
      id: 'homework',
    },
    {
      name: 'Emploi du temps',
      href: '/coming-soon',
      icon: 'today',
      id: 'schedule',
    },
    {
      name: 'Notes',
      href: '/coming-soon',
      icon: 'school',
      id: 'grades',
    },
    {
      name: 'Sac',
      href: '/coming-soon',
      icon: 'work_outline',
      id: 'bag',
    },
    'separator',
    {
      name: 'Paramètres',
      href: '/settings',
      icon: 'settings',
      id: 'settings',
    },
    {
      name: 'Vos contributions',
      href: '/reports',
      icon: 'bug_report',
      id: 'reports',
    },
    {
      name: 'Signaler un bug',
      modal: 'add-report',
      icon: 'bug_report',
      id: 'new-report',
    },
  ],
})

export const getters = {
  textColor: (state, getters) => backgroundColor =>
    // TODO: handle var(--) ?
    /* returns the corresponding text color most visible
     * on backgroundColor: either 'black' or 'white'.
     */
    tinycolor(backgroundColor).isLight() ? 'black' : 'white',
  formatTime: (state, getters) => time => {
    if (time === null) return null
    return format(time, 'HH:mm')
  },
  drawerLinks: state =>
    state.links.filter(link => {
      if (link === 'separator') return true
      return !['new-report'].includes(link.id)
    }),
  sideRailLinks: state =>
    state.links.filter(link => {
      if (link === 'separator') return false
      return ['timeline', 'notes', 'homework', 'grades', 'new-report'].includes(
        link.id
      )
    }),
}

export const mutations = {
  UPDATE_TIME: (state, newTime) => {
    state.now = newTime
    state.tomorrow = addDays(newTime, 1)
  },
  UPDATE_LOCATION: (state, newLocation) => {
    state.location = newLocation
  },
}

export const actions = {
  async loadAll({ dispatch }) {
    await dispatch('settings/load')
    await dispatch('subjects/load')
    await dispatch('homework/load')
    await dispatch('grades/load')
    await dispatch('schedule/load')
    await dispatch('notes/load')
    await dispatch('learndata/load')
  },

  async nuxtServerInit({ dispatch }) {
    await dispatch('settings/load')
    await dispatch('subjects/load')
  },

  acquireLocation({ commit, state }) {
    if (state.longitude === null || state.latitude === null)
      navigator.geolocation.getCurrentPosition(
        pos => {
          commit('UPDATE_LOCATION', pos.coords)
        },
        // eslint-disable-next-line
        (err) => {
          // eslint-disable-next-line
          this.$toast.error("Impossible d'obtenir la localisation")
        }
      )
  },
}

export const getValidator = ({
  constraints,
  resourceName,
  fieldNames,
  customConstraints = [],
  debug,
}) => getters => object => {
  debug = debug || false
  if (process.env.NODE_ENV !== 'development') debug = false
  /* Factory to create a validator.
  Describe constraints on fields, error messages are generated automatically.
  */

  if (debug) {
    console.group(`[validator] validating resource "${resourceName.name}"`) // eslint-disable-line
    console.log(`Constraints:`) // eslint-disable-line
    console.log({ ...constraints, ...customConstraints }) // eslint-disable-line
    console.log(`Fields:`) // eslint-disable-line
    console.log(
      // eslint-disable-line
      Object.fromEntries(
        Object.keys(fieldNames).map(field => [field, object[field]])
      )
    )
  }

  // Article in french
  const article = (noun, feminine, indeterminate = false) => {
    const vowels = ['a', 'e', 'i', 'o', 'u', 'y']
    if (vowels.includes(noun[0]) && !indeterminate) return "l'"
    else if (feminine) return indeterminate ? 'une ' : 'la '
    else return indeterminate ? 'un ' : 'le '
  }

  // Resource name article & indeterminate article
  resourceName.article = article(
    resourceName.name,
    resourceName.gender === 'F',
    false
  )
  resourceName.indeterminateArticle = article(
    resourceName.name,
    resourceName.gender === 'F',
    true
  )

  // Uppercase first char
  const upperFirst = string => string[0].toUpperCase() + string.slice(1)

  // Checkers
  const checkers = {
    maximum: ({ arg, fieldName }) =>
      object[fieldName] === null ||
      (typeof object[fieldName] === 'number' && object[fieldName] <= arg),
    minimum: ({ arg, fieldName }) =>
      object[fieldName] === null ||
      (typeof object[fieldName] === 'number' && object[fieldName] >= arg),
    maxLength: ({ arg, fieldName }) =>
      typeof object[fieldName] === 'string'
        ? object[fieldName].length <= arg
        : true,
    required: ({ arg, fieldName }) =>
      object.hasOwnProperty(fieldName) && object[fieldName] !== null,
    notEmpty: ({ arg, fieldName }) =>
      object[fieldName] && object[fieldName].trim().length > 0,
    isAWeekType: ({ arg, fieldName }) =>
      ['Q1', 'Q2', 'BOTH'].includes(object[fieldName]),
    before: ({ arg, fieldName }) => isBefore(object[fieldName], object[arg]),
    isAColor: ({ arg, fieldName }) =>
      object[fieldName].match(/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/) !== null,
    isAnEmail: ({ arg, fieldName }) =>
      // regex source: https://emailregex.com/
      typeof object[fieldName] === 'string' &&
      object[fieldName].match(
        /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
      ) !== null,
    isOfType: ({ arg, fieldName }) =>
      // eslint-disable-next-line valid-typeof
      typeof object[fieldName] === arg,
  }
  const check = ({ errorName, fieldName, errorArg }) => {
    // Wrap checkers with object property existential check
    /* Special case for required, which checks if the property exist
       and returnes *false* instead.
    */
    if (errorName !== 'required' && !object.hasOwnProperty(fieldName)) {
      return true
    }
    return checkers[errorName]({ arg: errorArg, fieldName })
  }

  // Error messages
  const message = ({ errorName, errorArg, fieldName }) => {
    const { gender, name } = fieldNames[fieldName]
    const fieldIsFeminine = gender === 'F'
    const determinateArticle = article(name, fieldIsFeminine, false)
    const indeterminateArticle = article(name, fieldIsFeminine, true)
    const argNameWithArticle =
      article(
        fieldNames[fieldName].name,
        fieldNames[fieldName].gender === 'F',
        false
      ) + fieldNames[fieldName].name
    const fieldNameWithArticle = determinateArticle + name
    return upperFirst(
      {
        maximum: `${fieldNameWithArticle} ne doit pas dépasser ${errorArg}`,
        minimum: `${fieldNameWithArticle} doit être d'au moins ${errorArg}`,
        maxLength: `${fieldNameWithArticle} ne doit pas dépasser ${errorArg} caractère${
          errorArg === 1 ? '' : 's'
        }`,
        required: `Veuillez renseigner ${indeterminateArticle}${name}`,
        isAWeekType: `${fieldNameWithArticle} doit être paire, impaire ou les deux.`,
        before: `${fieldNameWithArticle} doit être avant ${argNameWithArticle}`,
        isAColor: `${fieldNameWithArticle} doit être une couleur au format hexadécimal. Exemple: #09f ou #0479cf`,
        notEmpty: `${fieldNameWithArticle} est requis${
          fieldIsFeminine ? 'e' : ''
        }`,
        isAnEmail:
          name === 'adresse email'
            ? `${fieldNameWithArticle} doit être valide`
            : `${fieldNameWithArticle} doit être une adresse email valide`,
      }[errorName]
    )
  }

  // Stores the errors
  const errorMessages = {}
  // Fill each field with an empty array
  Object.keys(fieldNames).forEach(field => {
    errorMessages[field] = []
  })
  let validated = true

  // Go through the constraints
  for (const errorName in constraints) {
    if (constraints.hasOwnProperty(errorName)) {
      const fieldsOrArgs = constraints[errorName]
      // No arguments, field names are passed directly
      // eslint-disable-next-line
      if (fieldsOrArgs instanceof Array) {
        fieldsOrArgs.forEach(fieldName => {
          if (!check({ errorName, fieldName, errorArg: null })) {
            // Some error occured, add the error message
            errorMessages[fieldName].push(
              message({ errorName, fieldName, errorArg: null })
            )
            validated = false
          }
        })
      }
      // Error arguments, multiple cases. eg: maximum
      else {
        for (const errorArg in fieldsOrArgs) {
          if (fieldsOrArgs.hasOwnProperty(errorArg)) {
            const fields = fieldsOrArgs[errorArg]
            fields.forEach(fieldName => {
              if (!check({ errorName, fieldName, errorArg })) {
                errorMessages[fieldName].push(
                  message({ errorName, fieldName, errorArg })
                )
                validated = false
              }
            })
          }
        }
      }
    }
  }

  // Go through custom constraints
  customConstraints.forEach(({ message, constraint, field }) => {
    if (!constraint(getters, object)) {
      field = field === null ? 'nonFieldErrors' : field
      if (errorMessages.hasOwnProperty(field)) {
        errorMessages[field].push(message)
      } else {
        errorMessages[field] = [message]
      }
      validated = false
    }
  })

  // Creates a single string that reports all of the errors.
  // Get the total errors count
  let flatErrorMessage = ''
  let errorsCount = 0
  for (const field in errorMessages) {
    if (errorMessages.hasOwnProperty(field)) {
      const errors = errorMessages[field]
      if (errors.length) {
        flatErrorMessage += errors.join(', ') + '\n'
        errorsCount += errors.length
      }
    }
  }

  // returns the object
  const ret = {
    validated,
    errors: errorMessages,
    message: flatErrorMessage,
    count: errorsCount,
  }
  if (debug) {
    if (ret.validated) {
      console.log('Validated.')
    } else {
      console.log('Errors:')
      Object.entries(ret.errors).forEach(([name, errors]) => {
        if (errors.length) {
          console.log(`    ${name}:`)
          console.log(errors)
        }
      })
    }
    console.groupEnd()
  }

  return ret
}

export const getMutations = (
  what,
  mapWith = o => o,
  pure = true,
  verbs = ['add', 'set', 'del', 'patch'],
  primaryKey = 'uuid',
  debug = false
) => {
  /* Factory to create basic mutations while preserving DRYness of code
   * set, add, del & patch are booleans that—when set to false—disable
   * a particular mutation from being built & returned.
   *
   * @param pure: makes the function return solely the verb as the function name.
   *              useful for store modules consisting of *one* resource only.
   *              example: {SET:..., ADD:... ...} instead of {SET_THINGS:..., ADD_THING:... ...}
   *
   * @param what: the name of the resource. Will be used to calculate the
   *              state's array & mutations' names.
   * @param mapWith: a function to map each object to when putting into the state
   *                 Defaults to `o => o`
   * @param primaryKey: the name of the primary key used for mutations acting on a single object.
   * @param debug: Log every mutation if true
   */

  const mutations = {}
  const WHAT = pure ? '' : '_' + constantCase(what)
  const whats = what + 's'

  if (verbs.includes('set'))
    mutations[`SET${WHAT}${pure ? '' : 'S'}`] = (state, items) => {
      if (debug) {
        console.group(`${whats}/SET`)
        console.log('before:')
        console.log(state[whats])
      }
      Vue.set(state, whats, items.map(mapWith))
      if (debug) {
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }
  if (verbs.includes('add'))
    mutations[`ADD${WHAT}`] = (state, item) => {
      if (debug) {
        console.group(`${whats}/ADD`)
        console.log('item:')
        console.log(item)
        console.log('before:')
        console.log(state[whats])
      }
      Vue.set(state, whats, [...state[whats], mapWith(item)])
      if (debug) {
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }
  if (verbs.includes('del'))
    mutations[`DEL${WHAT}`] = (state, pk) => {
      if (debug) {
        console.group(`${whats}/DEL`)
        console.log('item:')
        console.log(state[whats][pk])
        console.log('before:')
        console.log(state[whats])
      }
      Vue.set(
        state,
        whats,
        state[whats].filter(o => o[primaryKey] !== pk)
      )
      if (debug) {
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }
  if (verbs.includes('patch'))
    mutations[`PATCH${WHAT}`] = (state, { pk, modifications }) => {
      if (debug) {
        console.group(`${whats}/PATCH`)
        console.log('before:')
        console.log(state[whats])
      }
      // Get the requested item's index in the state array
      const idx = state[whats].map(o => o[primaryKey]).indexOf(pk)
      // Apply modifications
      let item = { ...state[whats][idx], ...modifications }
      // Re-run mapWith
      item = mapWith(item)
      // Set in store
      Vue.set(state[whats], idx, item)
      if (debug) {
        console.log('item:')
        console.log(item)
        console.log('after:')
        console.log(state[whats])
        console.groupEnd()
      }
    }

  return mutations
}

export const removeByProp = (arrayOfObjects, propName, propValue) => {
  const arr = [...arrayOfObjects]
  return arr.splice(
    arr.findIndex(obj => obj[propName] === propValue),
    1
  )
}

// Reset time component to 00:00:00
// Useful to do date-wise comparison, w/o taking time into account.
export const removeTime = date => {
  date.setHours(0)
  date.setMinutes(0)
  date.setSeconds(0)
  return date
}
