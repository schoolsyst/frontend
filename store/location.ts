import { ActionTree, MutationTree } from 'vuex'
import { RootState } from '~/store'

export const state: () => {
  latitude: number | null
  longitude: number | null
} = () => ({
  latitude: null,
  longitude: null,
})

export type LocationState = ReturnType<typeof state>

export const actions: ActionTree<LocationState, RootState> = {
  acquireLocation({ commit, state }) {
    if (state.longitude === null || state.latitude === null)
      navigator.geolocation.getCurrentPosition(
        pos => {
          commit('UPDATE_LOCATION', pos.coords)
        },
        // eslint-disable-next-line
        err => {
          // eslint-disable-next-line
          this.$toast.error("Impossible d'obtenir la localisation")
        }
      )
  },
}

export const mutations: MutationTree<LocationState> = {
  UPDATE_LOCATION: (state, newLocation: typeof state) => {
    state = { ...state, ...newLocation }
  },
}
