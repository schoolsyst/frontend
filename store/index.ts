import { toDate, addDays } from 'date-fns'
import { getAccessorType } from 'nuxt-typed-vuex'
import { GetterTree, ActionTree, MutationTree } from 'vuex'

import * as homework from '~/store/homework'
import * as auth from '~/store/auth'
import * as bag from '~/store/bag'
import * as grades from '~/store/grades'
import * as learndata from '~/store/learndata'
import * as location from '~/store/location'
import * as navigation from '~/store/navigation'
import * as notes from '~/store/notes'
import * as reports from '~/store/reports'
import * as settings from '~/store/settings'
import * as subjects from '~/store/subjects'
import * as theme from '~/store/theme'
import * as schedule from '~/store/schedule'

export const state: () => {
  now: Date
  tomorrow: Date
  today: Date
} = () => ({
  now: toDate(Date.now()), // For time-dependent getters.
  tomorrow: addDays(toDate(Date.now()), 1),
  today: new Date(),
})

export type RootState = ReturnType<typeof state>

export const mutations: MutationTree<RootState> = {
  UPDATE_TIME: (state, newTime) => {
    state.now = newTime
    state.tomorrow = addDays(newTime, 1)
  },
}

export const actions: ActionTree<RootState, RootState> = {
  async loadAll({ dispatch }) {
    await dispatch('settings/load')
    await dispatch('subjects/load')
    await dispatch('homework/load')
    await dispatch('grades/load')
    await dispatch('schedule/load')
    await dispatch('notes/load')
    await dispatch('learndata/load')
  },

  async nuxtServerInit({ dispatch }) {
    await dispatch('settings/load')
    await dispatch('subjects/load')
  },
}

export const accessorType = getAccessorType({
  state,
  mutations,
  actions,
  modules: {
    homework,
    auth,
    bag,
    grades,
    learndata,
    location,
    navigation,
    notes,
    reports,
    schedule,
    settings,
    subjects,
    theme,
  },
})
