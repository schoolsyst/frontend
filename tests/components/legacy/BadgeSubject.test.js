import { mount, createLocalVue } from '@vue/test-utils'
import VTooltip from 'v-tooltip'
import Vuex from 'vuex'
import BadgeSubject from '../../../components/legacy/BadgeSubject.vue'
import { getters } from '../../../store/index'

const vue = createLocalVue()
vue.directive('tooltip', VTooltip)
vue.use(Vuex)
const store = {
  getters: { textColor: getters.textColor },
}
const mnt = (component, opts) =>
  mount(component, { ...opts, localVue: vue, store })

// TODO #77: Test w/ color:red and color:cyan

describe('<BadgeSubject>', () => {
  test('is a Vue instance', () => {
    const component = mnt(BadgeSubject)
    expect(component.isVueInstance()).toBeTruthy()
  })
  test('is a <button> when clickable', () => {
    const component = mnt(BadgeSubject, {
      propsData: {
        clickable: true,
      },
    })
    expect(component.is('button')).toBeTruthy()
  })
  test('is a <span> when not clickable', () => {
    const component = mnt(BadgeSubject)
    expect(component.is('span')).toBeTruthy()
  })
  describe("variant 'dot'", () => {
    test('matches snapshot', () => {
      const component = mnt(BadgeSubject, { propsData: { variant: 'dot' } })
      expect(component).toMatchSnapshot()
    })
  })
  describe("variant 'badge'", () => {
    test('matches snapshot', () => {
      const component = mnt(BadgeSubject, { propsData: { variant: 'badge' } })
      expect(component).toMatchSnapshot()
    })
  })
  test('click event is fired', () => {
    const component = mnt(BadgeSubject)
    component.trigger('click')
    expect(component.emitted('click')).toBeTruthy()
  })

  test('no tooltip is shown when has attribute [no-tooltip]', () => {
    const component = mnt(BadgeSubject, { propsData: { noTooltip: true } })
    expect(component.vm.tooltipContent).toBeNull()
  })
})
