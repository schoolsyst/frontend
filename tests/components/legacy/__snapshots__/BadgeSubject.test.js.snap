// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BadgeSubject> variant 'badge' matches snapshot 1`] = `<span data-variant="badge" type="button" class="subject"><i class="unknown-icon material-icons-outlined">more_horiz</i><span class="name">Choisir...</span></span>`;

exports[`<BadgeSubject> variant 'dot' matches snapshot 1`] = `<span data-variant="dot" type="button" class="subject"><i class="unknown-icon material-icons-outlined">more_horiz</i><span class="name">Choisir...</span></span>`;
